flake: TestMinorVersionCompatibility fails on transient ghcr.io image pull timeout

## Summary

`TestCelestiaTestSuite/TestMinorVersionCompatibility/v6_minor_versions` failed on `main` when pulling `ghcr.io/celestiaorg/celestia-app:v6.0.2-arabica`. The test currently aborts on any transient `ghcr.io` pull error (context deadline exceeded while awaiting headers) with no retry, which makes the job flaky on every post-merge CI run that catches ghcr.io having a bad moment.

## CI

- Post-merge CI run: https://github.com/celestiaorg/celestia-app/actions/runs/24667508191/job/72129669852
- Commit: `82e365a3f1ec6fb4175319d752425f1bc6745c60` (on `main`)

## Failure output

```
=== RUN   TestCelestiaTestSuite/TestMinorVersionCompatibility/v6_minor_versions
    e2e_minor_version_compatibility_test.go:73: Testing compatibility between versions: [v6.0.0-arabica v6.0.1-arabica v6.0.2-arabica v6.0.3-arabica v6.0.4-arabica v6.0.5-arabica]
    ...
    logger.go:146: 2026-04-20T13:00:32.338Z	ERROR	failed to run command	... "error": "job ghcr.io/celestiaorg/celestia-app:v6.0.2-arabica: pull image ghcr.io/celestiaorg/celestia-app:v6.0.2-arabica: Error response from daemon: Head \"https://ghcr.io/v2/celestiaorg/celestia-app/manifests/v6.0.2-arabica\": Get \"https://ghcr.io/token?scope=repository%3Acelestiaorg%2Fcelestia-app%3Apull&service=ghcr.io\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
    e2e_minor_version_compatibility_test.go:82:
        Error Trace: .../e2e_minor_version_compatibility_test.go:82
        Error:       Received unexpected error:
                     job ghcr.io/celestiaorg/celestia-app:v6.0.2-arabica: pull image ...: context deadline exceeded
        Messages:    failed to start mixed version chain
    e2e_minor_version_compatibility_test.go:78: Error stopping chain: failed to stop container: invalid container name or ID: value is empty
--- FAIL: TestCelestiaTestSuite (46.53s)
    --- FAIL: TestCelestiaTestSuite/TestMinorVersionCompatibility (46.14s)
        --- FAIL: TestCelestiaTestSuite/TestMinorVersionCompatibility/v6_minor_versions (46.14s)
```

## Root cause

The test at `test/docker-e2e/e2e_minor_version_compatibility_test.go:82` calls `chain.Start(ctx)` which triggers image pulls via tastora/docker. The test pulls images one per validator (6 images) and treats any pull error as fatal. When ghcr.io has a transient timeout (Docker daemon's client-side deadline expires awaiting response headers), the whole chain build aborts and the test fails.

A secondary symptom: the `t.Cleanup` handler then tries to stop a chain that never started, producing the noisy `Error stopping chain: failed to stop container: invalid container name or ID: value is empty` log.

## Proposed fix

1. Pre-pull the required images with retry+backoff before `buildMixedVersionChain` to separate registry flakiness from test logic; OR
2. Configure the tastora image-pull path to retry on context-deadline-exceeded errors; OR
3. Mark the test as retry-eligible at the CI level (`go test -count=1 -retries=N` / a rerun-failed step) while root cause is being addressed.

Option 1 is cleanest and keeps the failure signal distinct (infra vs. test).

--- a/test/docker-e2e/e2e_minor_version_compatibility_test.go
+++ b/test/docker-e2e/e2e_minor_version_compatibility_test.go
@@ -98,6 +98,14 @@ func (s *CelestiaTestSuite) buildMixedVersionChain(ctx context.Context, versionT
 	baseTag := versionTags[0]
 	baseCfg := dockerchain.DefaultConfig(s.client, s.network).WithTag(baseTag)
 
+	// Pre-pull all per-tag images with retry so that transient ghcr.io errors
+	// surface as infra noise rather than test failures. PullImage is idempotent
+	// and becomes a no-op once the image is cached locally.
+	for _, tag := range versionTags {
+		img := tastoracontainertypes.NewImage(baseCfg.Image, tag, "10001:10001")
+		s.Require().NoError(pullImageWithRetry(ctx, s.client, img), "failed to pre-pull image %s", img.Ref())
+	}
+
 	validators := make([]genesis.Validator, len(versionTags))
 	for i := range len(versionTags) {
 		validators[i] = genesis.NewDefaultValidator(fmt.Sprintf("validator%d", i))

--- a/test/docker-e2e/image_pull_retry.go
+++ b/test/docker-e2e/image_pull_retry.go
@@ -1,0 +1,45 @@
+package docker_e2e
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	tastoracontainertypes "github.com/celestiaorg/tastora/framework/docker/container"
+	tastoratypes "github.com/celestiaorg/tastora/framework/types"
+)
+
+// retryPull invokes op up to maxAttempts times, sleeping baseDelay*2^attempt
+// between attempts, and aborts early if ctx is cancelled. Returns the final
+// op error on exhaustion.
+func retryPull(ctx context.Context, maxAttempts int, baseDelay time.Duration, op func() error) error {
+	var lastErr error
+	for attempt := 0; attempt < maxAttempts; attempt++ {
+		if err := ctx.Err(); err != nil {
+			return fmt.Errorf("retryPull: %w", err)
+		}
+		lastErr = op()
+		if lastErr == nil {
+			return nil
+		}
+		if attempt == maxAttempts-1 {
+			break
+		}
+		delay := baseDelay << attempt
+		select {
+		case <-time.After(delay):
+		case <-ctx.Done():
+			return fmt.Errorf("retryPull: %w", ctx.Err())
+		}
+	}
+	return fmt.Errorf("retryPull exhausted after %d attempts: %w", maxAttempts, lastErr)
+}
+
+// pullImageWithRetry pulls the given image via tastora's idempotent PullImage,
+// retrying on transient registry errors (e.g. ghcr.io timeouts). Safe to call
+// on images that are already cached locally — PullImage is a no-op in that case.
+func pullImageWithRetry(ctx context.Context, client tastoratypes.TastoraDockerClient, image tastoracontainertypes.Image) error {
+	return retryPull(ctx, 3, 2*time.Second, func() error {
+		return image.PullImage(ctx, client)
+	})
+}

--- a/test/docker-e2e/image_pull_retry_test.go
+++ b/test/docker-e2e/image_pull_retry_test.go
@@ -1,0 +1,66 @@
+package docker_e2e
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRetryPull_SucceedsOnFirstAttempt(t *testing.T) {
+	calls := 0
+	err := retryPull(context.Background(), 3, time.Millisecond, func() error {
+		calls++
+		return nil
+	})
+
+	require.NoError(t, err)
+	assert.Equal(t, 1, calls)
+}
+
+func TestRetryPull_SucceedsAfterTransientFailures(t *testing.T) {
+	calls := 0
+	err := retryPull(context.Background(), 3, time.Millisecond, func() error {
+		calls++
+		if calls < 3 {
+			return fmt.Errorf("transient error %d", calls)
+		}
+		return nil
+	})
+
+	require.NoError(t, err)
+	assert.Equal(t, 3, calls)
+}
+
+func TestRetryPull_FailsAfterExhaustion(t *testing.T) {
+	calls := 0
+	sentinel := errors.New("ghcr boom")
+	err := retryPull(context.Background(), 3, time.Millisecond, func() error {
+		calls++
+		return sentinel
+	})
+
+	require.Error(t, err)
+	assert.ErrorIs(t, err, sentinel, "final error should wrap the last op error")
+	assert.Equal(t, 3, calls)
+}
+
+func TestRetryPull_RespectsContextCancellation(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	calls := 0
+	err := retryPull(ctx, 5, 10*time.Second, func() error {
+		calls++
+		if calls == 1 {
+			cancel()
+		}
+		return errors.New("boom")
+	})
+
+	require.Error(t, err)
+	assert.ErrorIs(t, err, context.Canceled)
+	assert.Equal(t, 1, calls, "should not invoke op again after ctx cancellation")
+}

